### PR TITLE
#77-create-main-operation-with-categories-via-mediatr

### PR DESCRIPTION
**Pull request by otash8110**

### CODE REVIEW

File: src/BudgetManager/BudgetManager.API/Controllers/CategoryController.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            var response = await _mediator.Send(new UpdateCategoryCommand(category), cancellationToken);
+            return Ok(response);
+        }
+
+        [HttpDelete("[action]/{id}")]
+        public async Task<IActionResult> DeleteOne(Guid id, CancellationToken cancellationToken)
+        {
+            var response = _mediator.Send(new DeleteCategoryCommand(id), cancellationToken);
+            return Ok(response);
```
Comment (Saigunas) [inline src/BudgetManager/BudgetManager.API/Controllers/CategoryController.cs]: If category deletion fails it is going to return ok response.
image
Should probably return a bad request.
Similar situation with some other controller operations.

Comment (otash8110) [reply src/BudgetManager/BudgetManager.API/Controllers/CategoryController.cs]: > If category deletion fails it is going to return ok response. image Should probably return a bad request. Similar situation with some other controller operations.

Will fix it

File: src/BudgetManager/BudgetManager.API/Controllers/CategoryController.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            var response = await _mediator.Send(new GetOneCategoryQuery(id), cancellationToken);
+            return response == null ? NotFound() : Ok(response);
+        }
+
+        [HttpPost("[action]")]
+        public async Task<IActionResult> InsertOne(AddCategoryDTO category, CancellationToken cancellationToken)
+        {
+            var response = await _mediator.Send(new AddCategoryCommand(category), cancellationToken);
+            return response == null ? BadRequest() : Ok(response);
```
Comment (VladislavGromadtsov) [inline src/BudgetManager/BudgetManager.API/Controllers/CategoryController.cs:37]: Can your response be null?

File: src/BudgetManager/BudgetManager.CQRS/Handlers/CategoryHandlers/AddCategoryHandler.cs [hunk 3]
```diff
@@ -1,22 +1,29 @@
...
         }
 
-        public Task<CategoryResponse> Handle(AddCategoryCommand request, CancellationToken cancellationToken)
+        public async Task<string> Handle(AddCategoryCommand request, CancellationToken cancellationToken)
         {
-            throw new NotImplementedException();
+            AddCategoryDTO requestCategory = request.category;
+            Category mappedCategory = _mapper.Map<Category>(requestCategory);
+            await _categoryRepository.InsertOneAsync(mappedCategory, cancellationToken);
+            return mappedCategory.Id.ToString();
```
Comment (VladislavGromadtsov) [inline src/BudgetManager/BudgetManager.CQRS/Handlers/CategoryHandlers/AddCategoryHandler.cs]: When inserting is not successfully done mappedCategory.Id will have Guid.Empty 00000000-0000-0000-0000-000000000000

**Review by VladislavGromadtsov — CHANGES_REQUESTED** [review]

**Review by VladislavGromadtsov — APPROVED** [review]

File: src/BudgetManager/BudgetManager.CQRS/Handlers/CategoryHandlers/GetCategoriesHandler.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            var allCategoriesFromDb = await _categoryRepository.GetAllAsync(cancellationToken);
+            var listOfResponseCategories = new List<CategoryResponse>();
+            foreach (var category in allCategoriesFromDb)
+            {
+                CategoryResponse mappedCategory = _mapper.Map<CategoryResponse>(category);
+                listOfResponseCategories.Add(mappedCategory);
+            }
+
+            return listOfResponseCategories;
```
Comment (Saigunas) [inline src/BudgetManager/BudgetManager.CQRS/Handlers/CategoryHandlers/GetCategoriesHandler.cs]: I think if we receive IEnumerable from repo we should also return it as IEnumerable.
Also this code could be shortened to one line of code.
Screenshot 2022-07-26 at 13-40-32 #77-create-main-operation-with-categories-via-mediatr by otash8110 · Pull Request #35 · MaximPyatigor_exsb22-t1

File: src/BudgetManager/BudgetManager.CQRS/Handlers/CategoryHandlers/DeleteCategoryHandler.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        public DeleteCategoryHandler(IBaseRepository<Category> categoryRepository)
+        {
+            _categoryRepository = categoryRepository;
+        }
+
+        public async Task<bool> Handle(DeleteCategoryCommand request, CancellationToken cancellationToken)
+        {
+            var idToDelete = request.id;
+
+            var result = await _categoryRepository.DeleteByIdAsync(idToDelete, cancellationToken);
```
Comment (Saigunas) [inline src/BudgetManager/BudgetManager.CQRS/Handlers/CategoryHandlers/DeleteCategoryHandler.cs]: could be shortened.

File: src/BudgetManager/BudgetManager.CQRS/Handlers/CategoryHandlers/GetCategoriesHandler.cs (hunk 4 above)
Comment (Saigunas) [reply src/BudgetManager/BudgetManager.CQRS/Handlers/CategoryHandlers/GetCategoriesHandler.cs]: Sorry for a sc. Github comments didn't want to accept this piece of text for some reason.

**Review by Saigunas — APPROVED** [review]

**Pull request merged**